fairscale/nn/data_parallel/sharded_ddp.py

```diff
@@ -4,234 +4,266 @@
+from fairscale.optim.utils import Workhandle
 
 
 class ShardedDataParallel(nn.Module):
-    """Implements distributed data parallel training with optimizer state sharding.
-
-    A simplified version of :class:`torch.nn.parallel.DistributedDataParallel`.
-    This version uses a c10d process group for communication and optionally
-    broadcast buffers.
-
-    Args:
-        module (~torch.nn.Module): module to be parallelized
-        optimizer (~torch.optim.Optimizer): optimizer to be used for training
-        optimizer_params(Dict): extra parameters for the optimizer
-        world_size (int): number of parallel workers
-        broadcast_buffers (bool): flag that enables syncing (broadcasting) buffers of
-        the module at beginning of the forward function. (default: ``True``)
-        process_group (optional): the c10d process group to be used for
-            distributed gradient reduction. If None, the default WORLD process group
-            will be used.
-        buffer_size (int, optional): number of elements to buffer before
-            performing reduce (default: 512k). Used to reduce multiple small
-            params to avoid communication overhead.
+    """
+    Wrap the model, and reduce the gradients to the right rank during the backward pass.
+
+    - the partition is given by the sharded optimizer
+    - wrap the base model with a model which knows where to reduce each gradient
+    - add an autograd function which calls the model grad dispatch on the way back
+
+     Args:
+        base_model (nn.Module):
+            model to be wrapped
+        sharded_optimizer (OSS, or list of OSS):
+            the sharded optimizer(s) which will decide the gradient partitioning
+
+    Keyword Args:
+        process_group (torch.nn.Optimizer):
+            optimizer to shard (default: SGD)
+        process_group (group):
+            torch.distributed group (default: group.WORLD)
+        broadcast_buffers (bool):
+            whether to broadcast model buffers in between ranks at the beginning of each forward pass
+        reduce_buffer_size (int):
+            the size of the per-device-per-rank buffers used for the reduce operation
+
     """
 
     def __init__(
         self,
-        module: nn.Module,
-        optimizer: Type[torch.optim.Optimizer],
-        optimizer_params: Dict[str, Any],
-        world_size: int,
-        broadcast_buffers: bool,
+        base_model: nn.Module,
+        sharded_optimizer: Union[OSS, List[OSS]],
         process_group: Any = None,
-        buffer_size: int = 2 ** 19,
+        broadcast_buffers: bool = True,
     ):
         super().__init__()
 
-        self.module = module
-        self.world_size = world_size
+        self.base_model = base_model
+        self.sharded_optimizers = [sharded_optimizer] if isinstance(sharded_optimizer, OSS) else sharded_optimizer
+        self.enable_broadcast_buffers = broadcast_buffers
+        self.accumulate_grads = False
```

Comment:
This is confusing out of context - maybe note that this will be temporarily changed in a context manager on `no_sync`?

Reply:
I added a comment and renamed to show that it's a bool